### PR TITLE
Fix check deletion in anti-entropy sync

### PR DESCRIPTION
**Pull request by freddygv**
When the anti-entropy sync calls `deleteService` we also need to delete all associated checks.

Finding the checks to delete requires looping over the `map[structs.CheckID]*CheckState` which contains checks for all services registered with the agent. While looping we need to see if the service ID of the check matches the service ID that we are deleting.

Currently we are only matching on the service ID string, and not the `structs.ServiceID`, which incorporates `EnterpriseMeta`. 

That means that if we have service ID `web` registered in namespaces `prod` and `dev`, then a call to delete `dev/web` would delete checks for `web` across both namespaces.

### CODE REVIEW

File: agent/local/state.go [hunk 1]
```diff
@@ -1098,8 +1098,11 @@ func (l *State) deleteService(key structs.ServiceID) error {
 		delete(l.services, key)
 		// service deregister also deletes associated checks
 		for _, c := range l.checks {
-			if c.Deleted && c.Check != nil && c.Check.ServiceID == key.ID {
-				l.pruneCheck(c.Check.CompoundCheckID())
+			if c.Deleted && c.Check != nil {
+				sid := c.Check.CompoundServiceID()
+				if sid.Matches(&key) {
+					l.pruneCheck(c.Check.CompoundCheckID())
+				}
```
Comment (freddygv) [inline agent/local/state.go:1105]: Could also do:

```suggestion
				if sid := c.Check.CompoundServiceID(); sid.Matches(&key) {
					l.pruneCheck(c.Check.CompoundCheckID())
				}
```

Comment (i0rek) [reply agent/local/state.go:1105]: I prefer the original version because it is more readable. I am used to the one-liner when used with `ok` and `err`, but here it doesn't improve the situation imho.

**Comment by freddygv** [comment]
Sounds good, I'll add a test in Enterprise.

**Comment by freddygv** [comment]
@i0rek Actually this issue was more subtle than I thought.

When we call `agent.RemoveService`, it calls `state.RemoveServiceWithChecks` with all the checks for that service. So in the service removal case: for the check to have been marked for deletion it must have been associated with the service/namespace.

That means that checks for the same service name in another namespace would not be pruned because they are not marked as Deleted.

However, if you call RemoveService on `dev/web` and then RemoveCheck on `prod/web/web-check`:
- `dev/web` and its checks get marked as deleted
- `prod/web/web-check` gets marked as deleted
- SyncChanges will first remove service `dev/web`:
  - `dev/web` gets deregistered from catalog and deleted from state
  - Then we loop over all checks and prune both its check and the one from prod/web (because we were only matching on ID)
 
That means that when SyncChanges goes to deregister `prod/web/web-check`, it can't, because it's already been deleted from the local state.

However, the next run to SyncFull will see that the check is present in the catalog but not in the local state. That will mark it for deregistration.

In summary: The issue from this PR is self-healing. The end result is that the removal of a check can sometimes take longer than expected. Likely no more than an additional minute.

**Comment by i0rek** [comment]
Thanks for sharing!
